src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php

```diff
@@ -170,11 +181,51 @@ public function process(ContainerBuilder $container)
         }
     }
 
+    /**
+     * Finds a list of alternative class names similar to a missing class.
+     *
+     * @param ContainerBuilder $container    The container builder instance
+     * @param string           $missingClass The missing class name to match
+     *
+     * @return string[] A list of up to `MAX_ALTERNATIVE_SUGGESTIONS` class names sorted by similarity
+     */
+    private function findAlternativeClassNames(ContainerBuilder $container, string $missingClass): array
+    {
+        $distances = [];
+        foreach ($container->getDefinitions() as $knownDefinition) {
+            $knownClass = $container->getParameterBag()->resolveValue($knownDefinition->getClass());
+
+            if (null !== $knownClass && $knownClass !== $missingClass) {
+                $distance = levenshtein($missingClass, $knownClass);
+                if($distance <= self::MAX_LEVENSHTEIN_DISTANCE) {
+                    $distances[$knownClass] = $distance;
+                }
+            }
+        }
+
+        asort($distances);
+
+        return array_slice(array_keys($distances), 0, self::MAX_ALTERNATIVE_SUGGESTIONS);
+    }
+
     private function getEventFromTypeDeclaration(ContainerBuilder $container, string $id, string $method): string
     {
+        $class = $container->getDefinition($id)->getClass();
+        $resolvedClass = $container->getParameterBag()->resolveValue($class);
+        if (!class_exists($resolvedClass)) {
+            $message = sprintf('Class "%s" for service "%s" does not exist.', $resolvedClass, $id);
+
+            $alternatives = $this->findAlternativeClassNames($container, $resolvedClass);
+
+            if (!empty($alternatives)) {
+                $message .= sprintf(' Did you mean %s?', implode(' or ', array_map(fn ($alt) => '"' . $alt . '"', $alternatives)));
+            }
+
+            throw new InvalidArgumentException($message);
+        }
+
         if (
-            null === ($class = $container->getDefinition($id)->getClass())
-            || !($r = $container->getReflectionClass($class, false))
+            !($r = $container->getReflectionClass($class, false))
```

Comment:
The bug is that the error message doesn't describe the real problem. What's needed is an error message indicating that the class doesn't exist.
It would be enough to leave the [exception here](https://github.com/symfony/symfony/blob/a2a47bc07b9997d663409fe0b97fa753a656036b/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L386):

```suggestion
            !($r = $container->getReflectionClass($class))
```